config.go

```diff
@@ -176,27 +187,62 @@ func (ca *Cagent) ReadConfigFromFile(configFilePath string, createIfNotExists bo
 		log.WithError(err).Errorf("Failed to create the config dir: '%s'", dir)
 	}
 
-	if _, err := os.Stat(configFilePath); os.IsNotExist(err) {
-		if !createIfNotExists {
-			return fmt.Errorf("Config file not exists: %s", configFilePath)
+	_, configExists := os.Stat(configFilePath)
+
+	var f *os.File
+	if f, err = os.OpenFile(configFilePath, os.O_RDWR|os.O_CREATE, 0644); err != nil {
+		// err is *os.PathError type and has all information about path and failure
+		return err
+	}
+
+	defer func() {
+		if err = f.Close(); err != nil {
+			log.WithError(err).Errorf("failed to close config file %s", configFilePath)
 		}
-		f, err := os.OpenFile(configFilePath, os.O_WRONLY|os.O_CREATE, 0644)
+	}()
 
-		if err != nil {
-			return fmt.Errorf("Failed to create the default config file: '%s'", configFilePath)
+	if os.IsNotExist(configExists) {
+		if !createIfNotExists {
+			return fmt.Errorf("config file not exists: %s", configFilePath)
 		}
-		defer f.Close()
-		enc := toml.NewEncoder(f)
-		enc.Encode(ca)
-	} else {
-		_, err = os.Stat(configFilePath)
-		if err != nil {
+
+		ca.lookupEnv()
+
+		var buf []byte
+		if buf, err = toml.Marshal(&MinValuableConfig{
+			HubURL:      ca.HubURL,
+			HubUser:     ca.HubUser,
+			HubPassword: ca.HubPassword,
+		}); err != nil {
 			return err
 		}
-		_, err = toml.DecodeFile(configFilePath, &ca)
+
+		writer := bufio.NewWriter(f)
+		if _, err = writer.Write(configAutogeneratedHeadline); err != nil {
+			log.Errorf("writing config %s + %s", configFilePath, err.Error())
+		} else {
+			if _, err = writer.Write(buf); err != nil {
+				log.Errorf("writing config %s + %s", configFilePath, err.Error())
+			}
+		}
+
+		if err = writer.Flush(); err != nil {
+			log.Errorf("writing config %s + %s", configFilePath, err.Error())
+		} else {
+			log.Infof("generated minimum valuable config: %s", configFilePath)
+		}
+	} else if configExists == nil {
+		// todo(troian) replace with toml.Unmarshal when https://github.com/pelletier/go-toml/issues/252 fixed
+		//  for now github.com/pelletier/go-toml completely overrides default values even if they omitted in the file
+
+		_, err = btoml.DecodeFile(configFilePath, &ca)
 		if err != nil {
 			return err
 		}
+
+		ca.lookupEnv()
+	} else {
+		return err
```

Comment:
what error are you returning here?

Reply:
In case err is neither nil nor NotFound

Reply:
This is another wild function.... so hard to read and follow the flow.

Reply:
Well, we can extend task to do refactoring

Reply:
Sounds good